lib/runtime/wazero/instance.go

```diff
@@ -0,0 +1,828 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package wazero_runtime
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/internal/log"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/crypto"
+	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
+	"github.com/ChainSafe/gossamer/lib/keystore"
+	"github.com/ChainSafe/gossamer/lib/runtime"
+	"github.com/ChainSafe/gossamer/lib/runtime/offchain"
+	"github.com/ChainSafe/gossamer/lib/transaction"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/klauspost/compress/zstd"
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/api"
+)
+
+// Name represents the name of the interpreter
+const Name = "wazero"
+
+type contextKey string
+
+const runtimeContextKey = contextKey("runtime.Context")
+
+var _ runtime.Instance = &Instance{}
+
+// Instance backed by wazero.Runtime
+type Instance struct {
+	Runtime wazero.Runtime
+	Module  api.Module
+	Context *runtime.Context
+	sync.Mutex
+}
+
+// Config is the configuration used to create a Wasmer runtime instance.
+type Config struct {
+	Storage     runtime.Storage
+	Keystore    *keystore.GlobalKeystore
+	LogLvl      log.Level
+	Role        common.NetworkRole
+	NodeStorage runtime.NodeStorage
+	Network     runtime.BasicNetwork
+	Transaction runtime.TransactionState
+	CodeHash    common.Hash
+}
+
+func decompressWasm(code []byte) ([]byte, error) {
+	compressionFlag := []byte{82, 188, 83, 118, 70, 219, 142, 5}
+	if !bytes.HasPrefix(code, compressionFlag) {
+		return code, nil
+	}
+
+	decoder, err := zstd.NewReader(nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating zstd reader: %s", err)
+	}
+	bytes, err := decoder.DecodeAll(code[len(compressionFlag):], nil)
+	if err != nil {
+		return nil, err
+	}
+	return bytes, err
+}
+
+// NewInstance instantiates a runtime from raw wasm bytecode
+func NewInstance(code []byte, cfg Config) (instance *Instance, err error) {
+	logger.Patch(log.SetLevel(cfg.LogLvl), log.SetCallerFunc(true))
+
+	ctx := context.Background()
+	rt := wazero.NewRuntime(ctx)
+
+	_, err = rt.NewHostModuleBuilder("env").
```

Comment:
Maybe a separate function or file for these?

Reply:
I'm going to leave it in for now.  It's a little long, but all the unimplemented exports are at the top, which are clear to see.